### Module 1 / Week4: Converting Unstructured Data to Features
### ---------------------------------------TOP----------------------------------------------
<li> DATA SCIENCE LSA LAB ASSIGNMENT.
<li> Rajesh Ashok Jaiswal UIN:669012329
<li> Datated:09/19/2016
#### ----------------------------------------

#### In this step will fetch data from Baseball group in to 'rajcorpus'

In [1]:
from sklearn.datasets import fetch_20newsgroups
categories = ['rec.sport.baseball']
dataset = fetch_20newsgroups(subset='all',shuffle=True, random_state=42, categories=categories)
rajcorpus = dataset.data #Data fetched from baseball group news into local data set 'rajcorpus'

In [2]:
#Data of rajcorpus
rajcorpus

["From: writingctr@leo.bsuvc.bsu.edu\nSubject: Re: CUB fever.\nOrganization: Ball State University, Muncie, In - Univ. Computing Svc's\nLines: 21\n\n\nIn article <kingoz.735285670@camelot>, kingoz@camelot.bradley.edu (Orin Roth) writes:\n> \n>    CUB fever is hitting me again. I'm beginning to think they have a \n>    chance this year. (what the heck am i thinking?)\n>    Sorry. Just a moment of incompetence.\n>    I'll be ok. Really. \n>    Orin.\n>    Bradley U.\n> \n> --\n> I'm really a jester in disguise!                                   \nI hear ya!  Then again, we must remember that we are indeed Cub fans, and\nthat the Cubs will eventually blow it.  After all, the Cubs are the easiest\nteam in the National League to root for.  No Pressure.  You know they will\nlose eventually.  Oh well, I suppose we must have faith.  After all, they\ndo look pretty good, and they don't even have Sandberg back yet.  \n\nCUBS IN '93!!!!!\n\nCHA\n",
 'From: schmke@cco.caltech.edu (Kevin Todd Schmi

#### In next few steps will collect reducndent word of html page and remove them

In [3]:
from bs4 import BeautifulSoup
import nltk
from nltk.corpus import stopwords
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.decomposition import TruncatedSVD

In [4]:
#Download stopward dictonary
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/rajeshjaiswal/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

#### In this step will update stopword file

In [6]:
stopset = set(stopwords.words('english'))
stopset.update(['lt','p','/p','br','amp','quot','field','font','normal','span','0px','rgb','style','51', 
                'spacing','text','helvetica','size','family', 'space', 'arial', 'height', 'indent', 'letter'
                'line','none','sans','serif','transform','line','variant','weight','times', 'new','strong', 'video', 'title'
                'white','word','letter', 'roman','0pt','16','color','12','14','21', 'neue', 'apple', 'class',  ])


### In this step we will work toward TF-IDF vectorizing

We are going to use scikit-learn's TF-IDF vectorizer to convert each document into a sparse matrix of TFIDF.

In [7]:
#Data of 1st document in corpus
rajcorpus[0]

"From: writingctr@leo.bsuvc.bsu.edu\nSubject: Re: CUB fever.\nOrganization: Ball State University, Muncie, In - Univ. Computing Svc's\nLines: 21\n\n\nIn article <kingoz.735285670@camelot>, kingoz@camelot.bradley.edu (Orin Roth) writes:\n> \n>    CUB fever is hitting me again. I'm beginning to think they have a \n>    chance this year. (what the heck am i thinking?)\n>    Sorry. Just a moment of incompetence.\n>    I'll be ok. Really. \n>    Orin.\n>    Bradley U.\n> \n> --\n> I'm really a jester in disguise!                                   \nI hear ya!  Then again, we must remember that we are indeed Cub fans, and\nthat the Cubs will eventually blow it.  After all, the Cubs are the easiest\nteam in the National League to root for.  No Pressure.  You know they will\nlose eventually.  Oh well, I suppose we must have faith.  After all, they\ndo look pretty good, and they don't even have Sandberg back yet.  \n\nCUBS IN '93!!!!!\n\nCHA\n"

### In this step will create sparse matrix

In [8]:
rajvectorizer = TfidfVectorizer(stop_words=stopset,
                                 use_idf=True, ngram_range=(1, 3))
rajX = rajvectorizer.fit_transform(rajcorpus)

In [10]:
#1st Documents matrix presentation will tell us number word it have
rajX[0]

<1x188088 sparse matrix of type '<class 'numpy.float64'>'
	with 229 stored elements in Compressed Sparse Row format>

In [11]:
#print sparse matrix 'rajX'
print (rajX[0])

  (0, 51441)	0.0739614450823
  (0, 187358)	0.0739614450823
  (0, 28939)	0.0739614450823
  (0, 145086)	0.0739614450823
  (0, 64046)	0.0739614450823
  (0, 77593)	0.0739614450823
  (0, 132368)	0.0739614450823
  (0, 102622)	0.0739614450823
  (0, 66435)	0.0739614450823
  (0, 113709)	0.0739614450823
  (0, 161841)	0.0739614450823
  (0, 179412)	0.0739614450823
  (0, 118825)	0.0739614450823
  (0, 64186)	0.0739614450823
  (0, 103114)	0.0739614450823
  (0, 94983)	0.0739614450823
  (0, 132283)	0.0739614450823
  (0, 142244)	0.0739614450823
  (0, 97593)	0.0739614450823
  (0, 114349)	0.0739614450823
  (0, 164212)	0.0739614450823
  (0, 59267)	0.0739614450823
  (0, 51481)	0.0739614450823
  (0, 35716)	0.0739614450823
  (0, 64172)	0.0739614450823
  :	:
  (0, 183911)	0.015655015889
  (0, 142480)	0.0555802309707
  (0, 121494)	0.111160461941
  (0, 37508)	0.0996785439776
  (0, 40682)	0.105257944488
  (0, 16584)	0.0739614450823
  (0, 94387)	0.111160461941
  (0, 25764)	0.0161922460089
  (0, 100274)	0.010320800

In [13]:
#Details of full rajX matrix with number of documents and total words count of rajcorpus.
rajX.shape

(994, 188088)

### In this step will apply LSA technique

Input: rajX, a matrix where m is the number of documents, and n is the number of word.
         Matrix rajX  m=994,n=188088 

Process: Applied  decomposition  on rajX and divided into three matricies called rajU, rajS, and rajV. 
When we do the decomposition, we have to pick a value k, that's how many concepts we are going to keep.

We are going to consider first 36 Concepts there for k=36.
<li>Hence,rajU[m,k]=rajU[994,36]
      <li>rajS[k,k]=rajS[36,36]
      <li>rajV[n,k]=rajV[188088,36] transpose matrix

In [14]:
#Considering concepts n=36
rajlsa = TruncatedSVD(n_components=36, n_iter=100)
rajlsa.fit(rajX)

TruncatedSVD(algorithm='randomized', n_components=36, n_iter=100,
       random_state=None, tol=0.0)

In [16]:
#1st row of component
rajlsa.components_[0]

array([ 0.01601322,  0.00499985,  0.00078314, ...,  0.00105261,
        0.00105261,  0.00105261])

In [17]:
#Will Check version of python and gcc 
import sys
print (sys.version)

3.5.2 |Anaconda 4.1.1 (x86_64)| (default, Jul  2 2016, 17:52:12) 
[GCC 4.2.1 Compatible Apple LLVM 4.2 (clang-425.0.28)]


## ------------------------------------------------------------------- 
### Final Step :To generate concepts report of first 36 rows

In [18]:
rajterms = rajvectorizer.get_feature_names()
for i, comp in enumerate(rajlsa.components_): 
    termsInComp = zip (rajterms,comp)
    sortedTerms =  sorted(termsInComp, key=lambda x: x[1], reverse=True) [:10]
    print ("Concept %d:" % i)
    for term in sortedTerms:
        print (term[0])
    print (" ")

Concept 0:
edu
com
year
writes
team
would
game
article
cs
baseball
 
Concept 1:
jewish
edu
subject
base
people
organization university
games
players
lafayette
lafibm
 
Concept 2:
year
com
get
ibm
time
last
ibm com
players
aix
really
 
Concept 3:
would
get
year
well
go
morris
clutch
sabo
00 00 baltimore
point
 
Concept 4:
hitter
batting
would
get
see
going
come
clutch
like
league
 
Concept 5:
00
better
gant
pitching
much
hit
big
00 00
edu organization
also
 
Concept 6:
00
cs
article
gant
pitcher
two
baseball
jewish
00 00 baltimore
make
 
Concept 7:
00
cs
00 00
best
fans
go
last year
many
first
last
 
Concept 8:
think
games
team
lines
last
still
also
anyone
years
one
 
Concept 9:
00
00 00
nntp
00 00 00
com
nntp posting
know
00 00 01
years
nntp posting host
 
Concept 10:
00
00 00
think
year
one
00 00 00
game
season
baseball
john
 
Concept 11:
would
team
one
first
host
two
players
know
00
usa
 
Concept 12:
first
clutch
nntp posting host
posting host
edu
nntp
really
still
subject
nntp posti

## Final report 

<li> From this analysis we can say that this data is about
<li> Baseball match National Leage
<li> People arround different university posting data.
<li> nntp posting host is a format for sending mail.
<h3>---------------------------------END of Report-----------------------------

### --------------------More evaluation work done on other group info--------------------------


In [20]:
# Here will work of sci.space dataset
from sklearn.datasets import fetch_20newsgroups
categories = ['sci.space']
dataset = fetch_20newsgroups(subset='all',shuffle=True, random_state=42, categories=categories)
rajspace = dataset.data #Data fetched from space group news into local data set 'rajspace'

In [21]:
#Detailed data of rajspace corpus
rajspace

['From: aws@iti.org (Allen W. Sherzer)\nSubject: Re: Orbital RepairStation\nOrganization: Evil Geniuses for a Better Tomorrow\nLines: 20\n\nIn article <C5HCBo.Joy@zoo.toronto.edu> henry@zoo.toronto.edu (Henry Spencer) writes:\n\n>The biggest problem with this is that all orbits are not alike.  It can\n>actually be more expensive to reach a satellite from another orbit than\n>from the ground.  \n\nBut with cheaper fuel from space based sources it will be cheaper to \nreach more orbits than from the ground.\n\nAlso remember, that the presence of a repair/supply facility adds value\nto the space around it. If you can put your satellite in an orbit where it\ncan be reached by a ready source of supply you can make it cheaper and gain\nbenefit from economies of scale.\n\n  Allen\n-- \n+---------------------------------------------------------------------------+\n| Lady Astor:   "Sir, if you were my husband I would poison your coffee!"   |\n| W. Churchill: "Madam, if you were my wife, I would

In [22]:
from bs4 import BeautifulSoup
import nltk
from nltk.corpus import stopwords
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.decomposition import TruncatedSVD

In [23]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/rajeshjaiswal/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [27]:
# Will add more word to avoid redundancy like this,is,space and NASA
stopset = set(stopwords.words('english'))
stopset.update(['lt','p','/p','br','amp','quot','field','font','normal','span','0px','rgb','style','51', 
                'spacing','text','helvetica','size','family', 'space', 'arial', 'height', 'indent', 'letter'
                'line','none','sans','serif','transform','line','variant','weight','times', 'new','strong', 'video', 'title'
                'white','word','letter', 'roman','0pt','16','color','12','14','21', 'neue', 'apple', 'class','NASA','this','space','$','-------'])


In [28]:
rajspace[0]

'From: aws@iti.org (Allen W. Sherzer)\nSubject: Re: Orbital RepairStation\nOrganization: Evil Geniuses for a Better Tomorrow\nLines: 20\n\nIn article <C5HCBo.Joy@zoo.toronto.edu> henry@zoo.toronto.edu (Henry Spencer) writes:\n\n>The biggest problem with this is that all orbits are not alike.  It can\n>actually be more expensive to reach a satellite from another orbit than\n>from the ground.  \n\nBut with cheaper fuel from space based sources it will be cheaper to \nreach more orbits than from the ground.\n\nAlso remember, that the presence of a repair/supply facility adds value\nto the space around it. If you can put your satellite in an orbit where it\ncan be reached by a ready source of supply you can make it cheaper and gain\nbenefit from economies of scale.\n\n  Allen\n-- \n+---------------------------------------------------------------------------+\n| Lady Astor:   "Sir, if you were my husband I would poison your coffee!"   |\n| W. Churchill: "Madam, if you were my wife, I would 

In [29]:
rajsvectorizer = TfidfVectorizer(stop_words=stopset,
                                 use_idf=True, ngram_range=(1, 3))
rajsX = rajsvectorizer.fit_transform(rajspace)

In [30]:
rajsX[0]

<1x254851 sparse matrix of type '<class 'numpy.float64'>'
	with 239 stored elements in Compressed Sparse Row format>

In [31]:
print (rajsX[0])

  (0, 90900)	0.0528748645448
  (0, 64824)	0.0528748645448
  (0, 13701)	0.0725866464284
  (0, 73794)	0.0725866464284
  (0, 250891)	0.0725866464284
  (0, 248237)	0.0540626550727
  (0, 136172)	0.0540626550727
  (0, 50162)	0.0540626550727
  (0, 52001)	0.0540626550727
  (0, 173876)	0.0540626550727
  (0, 251312)	0.0540626550727
  (0, 110871)	0.0540626550727
  (0, 207003)	0.0540626550727
  (0, 31777)	0.0540626550727
  (0, 124835)	0.0540626550727
  (0, 23177)	0.0547100719338
  (0, 198532)	0.0769166519558
  (0, 76028)	0.0769166519558
  (0, 38169)	0.0769166519558
  (0, 96801)	0.0769166519558
  (0, 49536)	0.0769166519558
  (0, 137433)	0.0769166519558
  (0, 220930)	0.0769166519558
  (0, 210944)	0.0769166519558
  (0, 186545)	0.0769166519558
  :	:
  (0, 39138)	0.0597292912743
  (0, 251946)	0.016239269138
  (0, 213040)	0.0336472044277
  (0, 107219)	0.0646758420702
  (0, 76346)	0.0289946929116
  (0, 232212)	0.0680505308336
  (0, 254775)	0.0692374036759
  (0, 120762)	0.0621122652744
  (0, 44317)	0.0769

In [32]:
rajsX.shape

(987, 254851)

In [34]:
rajslsa = TruncatedSVD(n_components=18, n_iter=100)
rajslsa.fit(rajsX)

TruncatedSVD(algorithm='randomized', n_components=18, n_iter=100,
       random_state=None, tol=0.0)

In [35]:
rajslsa.components_[0]

array([ 0.00875805,  0.00064517,  0.00020316, ...,  0.00207546,
        0.00207546,  0.00207546])

In [36]:
rajsterms = rajvectorizer.get_feature_names()
for i, comp in enumerate(rajslsa.components_): 
    termsInComp = zip (rajsterms,comp)
    sortedTerms =  sorted(termsInComp, key=lambda x: x[1], reverse=True) [:10]
    print ("Concept %d:" % i)
    for term in sortedTerms:
        print (term[0])
    print (" ")

Concept 0:
glory days born
matt williams
9834 blue
signals almost
99 snf
foolish get clue
data see
thomas hal morris
always memorable kinda
balls walks
 
Concept 1:
9834 blue
mention making
99 snf
foolish get clue
suppose seems favor
somebody example
garrison subject
les poe
sport baseball healey
sport baseball looked
 
Concept 2:
signals almost
data see
simmons 10
lloyd nebres uc
points first
suppose seems favor
balls walks
ha police exercise
les poe
randy johnson
 
Concept 3:
always memorable kinda
always pitching lacked
view ca distribution
see earlier game
stanky
00 00 00
view ca
view ca 94039
get good
posting made
 
Concept 4:
stretch led
might rushed lilliquist
see earlier game
matt williams
signals almost
prospect simple
pasadena lines 34
pitch pitch
funs
data see
 
Concept 5:
might rushed lilliquist
thomas hal morris
edge florida overall
points first
matt williams
city relief
00 00 01
reporting said say
talk phillies nntp
signals almost
 
Concept 6:
points first
suppose seems f

In [38]:
# 9834 blue
#mention making
#99 snf

### -------------------------------------End Of Assignment------------------------------